cpp/log/cluster_state_controller-inl.h

```diff
@@ -0,0 +1,203 @@
+#ifndef CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_INL_H_
+#define CERT_TRANS_LOG_CLUSTER_STATE_CONTROLLER_INL_H_
+
+#include "log/cluster_state_controller.h"
+#include "log/etcd_consistent_store-inl.h"
+
+#include <functional>
+
+namespace cert_trans {
+
+
+template <class Logged>
+ClusterStateController<Logged>::ClusterStateController(
+    util::Executor* executor, ConsistentStore<Logged>* store,
+    const MasterElection* election, const int min_serving_nodes,
+    const double min_serving_fraction)
+    : executor_(CHECK_NOTNULL(executor)),
+      store_(CHECK_NOTNULL(store)),
+      election_(CHECK_NOTNULL(election)),
+      min_serving_nodes_(min_serving_nodes),
+      min_serving_fraction_(min_serving_fraction),
+      exiting_(false),
+      update_required_(false),
+      cluster_serving_sth_update_thread_(
+          std::bind(&ClusterStateController<Logged>::ClusterServingSTHUpdater,
+                    this)) {
+  store_->WatchClusterNodeStates(
+      std::bind(&ClusterStateController::OnClusterStateUpdated, this,
+                std::placeholders::_1));
+}
+
+
+template <class Logged>
+ClusterStateController<Logged>::~ClusterStateController() {
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    exiting_ = true;
+    update_required_cv_.notify_all();
+  }
+  cluster_serving_sth_update_thread_.join();
+}
+
+
+template <class Logged>
+void ClusterStateController<Logged>::NewTreeHead(
+    const ct::SignedTreeHead& sth) {
+  CHECK(sth.IsInitialized());
```

Comment:
That's true, if any get added then it'll magically start working though - got any opinions on whether to leave it or ding it (I'm fairly easy either way)?